nilearn/region_extraction/regionextractor.py

```diff
@@ -0,0 +1,232 @@
+        assigned as each seperate region by
+        concatenating in the 4th dimension to a 4D image.
+    """
+    regions_img = []
+    affine = img.get_affine()
+    region = img_data.copy()
+
+    labels, n_labels = label(img_data)
+    labels_size = np.bincount(labels.ravel())
+    labels_size[0] = 0.
+
+    for label_id, label_size in enumerate(labels_size):
+        if label_size > min_size:
+            region = img_data.copy()
+            region = (labels == label_id) * img_data
+            region_img = new_img_like(img, region, affine)
+            regions_img.append(region_img)
+
+    return regions_img
+
+
+class RegionExtractor():
+    """ Region Extraction is a post processing technique which
+    is implemented to automatically extracts each brain atlas maps
+    into different set of seperated brain activated region.
+    Particularly, to show that each decomposed brain maps can be
+    used to focus on a target specific Regions of Interest analysis.
+    The idea to implement is a two step procedure:
+        1. Foreground Extraction using automatic thresholding function
+            based upon "percentile" of the features.
+            Ref: "fast_abs_percentile" in nilearn._utils.extmath
+        2. Biggest connected components extraction and assigning each
+            component a label to seperate out into each different region.
+            Ref: "label" in scipy.ndimage
+    Based upon the idea, two types of region extraction techniques
+    are designed:
+        1. "Auto Extractor" is a simple mechanism which is a way to
+            follows a two step procedure.
+        2. "Random Walker Extractor" in addition to a two step procedure
+            RWE is a way more robust mechanism than "Auto Extractor".
+            It goes this way which assigns labels using seed markers by
+            assigning each label with a seed point by using a standard
+            "peak_local_max" from scikit image.
+            Ref: "peak_local_max" in skimage.measure
+            Ref: "random_walker" in skimage.segmentation
+
+    Parameters
+    ----------
+    n_regions: int, default is None
+        limit the number of regions to extract from a set of brain maps.
+
+    min_size: int, default size 20, optional
+        restricts or suppresses the smallest spurious regions by selecting
+        the regions within which have voxels survived more than "min_size".
+
+    threshold: float or string, default string {'auto'}, optional
+        chooses to extract the foreground objects. If float, this value is
+        directly taken as a threshold or 'auto' option chooses automatically.
+        We recommend to use default which promises to extract reasonable
+        foreground objects.
+
+    percentile: int, range [0 100], default=80, optional
+        specifically used in threshold strategy which selects features
+        based on percentile value.
+
+    extractor: {'autoextractor', 'randomwalker'}, optional
```

Comment:
> The main difference between the two strategies is that random walker will try
> to separate connected components into subregions based on seeds. We have to
> find names reflecting this behavior because "random_walker" won't speak to the
> users. Maybe simple and seed_based.

'connect_components', and 'hill_detection'?